### PR TITLE
Extract re-usable modules from Longshot application.

### PR DESCRIPTION
**Pull request by DFurnes**
This pull request extracts re-usable Laravel Heroku app, MariaDB RDS instance, S3 bucket, SQS queue, and IAM user modules from the previously monolithic `longshot/application` module.

This shouldn't result in any actual changes to these resources, although it will take a bit of manual labor to migrate existing resources in Terraform's state to their new paths in each module, e.g:

```diff
- module.longshot.module.longshot-footlocker.aws_s3_bucket.storage
+ module.longshot.module.longshot-footlocker.module.storage.aws_s3_bucket.bucket
```

### CODE REVIEW

**Comment by DFurnes** [comment]
I've updated remote state to reflect the new modules in this pull request, so there are minimal actual changes to apply – just enabling New Relic on `footlocker-internal` since I've changed this to enable/disable automatically based on `environment` rather than a separate flag:

```
~ module.longshot.module.longshot-footlocker-internal.module.app.heroku_app.app
    config_vars.0.NEW_RELIC_APP_NAME:      "" => "longshot-footlocker-internal"
    config_vars.0.NEW_RELIC_ENABLED:       "false" => "true"
    config_vars.0.NEW_RELIC_LICENSE_KEY:   "" => "****** SECRET ******"
```

<details>
<summary>
state changes
</summary>

```sh
terraform state mv module.longshot.module.hrblock.aws_db_instance.database module.longshot.module.hrblock.module.database.aws_db_instance.database 
terraform state mv module.longshot.module.hrblock.aws_iam_access_key.aws_key module.longshot.module.hrblock.module.iam_user.aws_iam_access_key.aws_key
terraform state mv module.longshot.module.hrblock.aws_iam_user.aws_user module.longshot.module.hrblock.module.iam_user.aws_iam_user.aws_user
terraform state mv module.longshot.module.hrblock.aws_iam_user_policy.s3_policy module.longshot.module.hrblock.module.storage.aws_iam_user_policy.s3_policy
terraform state mv module.longshot.module.hrblock.aws_iam_user_policy.sqs_policy module.longshot.module.hrblock.module.queue.aws_iam_user_policy.sqs_policy
terraform state mv module.longshot.module.hrblock.aws_s3_bucket.storage module.longshot.module.hrblock.module.storage.aws_s3_bucket.bucket
terraform state mv module.longshot.module.hrblock.aws_sqs_queue.queue module.longshot.module.hrblock.module.queue.aws_sqs_queue.queue
terraform state mv module.longshot.module.hrblock.heroku_addon.redis module.longshot.module.hrblock.module.app.heroku_addon.redis
terraform state mv module.longshot.module.hrblock.heroku_app.app module.longshot.module.hrblock.module.app.heroku_app.app
terraform state mv module.longshot.module.hrblock.heroku_domain.domain module.longshot.module.hrblock.module.app.heroku_domain.domain
terraform state mv module.longshot.module.hrblock.heroku_drain.papertrail module.longshot.module.hrblock.module.app.heroku_drain.papertrail
terraform state mv module.longshot.module.hrblock.heroku_formation.queue module.longshot.module.hrblock.module.app.heroku_formation.queue
terraform state mv module.longshot.module.hrblock.heroku_formation.web module.longshot.module.hrblock.module.app.heroku_formation.web
terraform state mv module.longshot.module.hrblock.heroku_pipeline_coupling.app module.longshot.module.hrblock.module.app.heroku_pipeline_coupling.app
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_db_instance.database module.longshot.module.longshot-footlocker-internal.module.database.aws_db_instance.database
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_iam_access_key.aws_key module.longshot.module.longshot-footlocker-internal.module.iam_user.aws_iam_access_key.aws_key
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_iam_user.aws_user module.longshot.module.longshot-footlocker-internal.module.iam_user.aws_iam_user.aws_user
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_iam_user_policy.s3_policy module.longshot.module.longshot-footlocker-internal.module.storage.aws_iam_user_policy.s3_policy
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_iam_user_policy.sqs_policy module.longshot.module.longshot-footlocker-internal.module.queue.aws_iam_user_policy.sqs_policy
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_s3_bucket.storage module.longshot.module.longshot-footlocker-internal.module.storage.aws_s3_bucket.bucket
terraform state mv module.longshot.module.longshot-footlocker-internal.aws_sqs_queue.queue module.longshot.module.longshot-footlocker-internal.module.queue.aws_sqs_queue.queue
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_addon.redis module.longshot.module.longshot-footlocker-internal.module.app.heroku_addon.redis
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_app.app module.longshot.module.longshot-footlocker-internal.module.app.heroku_app.app
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_domain.domain module.longshot.module.longshot-footlocker-internal.module.app.heroku_domain.domain
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_drain.papertrail module.longshot.module.longshot-footlocker-internal.module.app.heroku_drain.papertrail
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_formation.queue module.longshot.module.longshot-footlocker-internal.module.app.heroku_formation.queue
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_formation.web module.longshot.module.longshot-footlocker-internal.module.app.heroku_formation.web
terraform state mv module.longshot.module.longshot-footlocker-internal.heroku_pipeline_coupling.app module.longshot.module.longshot-footlocker-internal.module.app.heroku_pipeline_coupling.app
terraform state mv module.longshot.module.longshot-footlocker.aws_db_instance.database module.longshot.module.longshot-footlocker.module.database.aws_db_instance.database
terraform state mv module.longshot.module.longshot-footlocker.aws_iam_access_key.aws_key module.longshot.module.longshot-footlocker.module.iam_user.aws_iam_access_key.aws_key
terraform state mv module.longshot.module.longshot-footlocker.aws_iam_user.aws_user module.longshot.module.longshot-footlocker.module.iam_user.aws_iam_user.aws_user
terraform state mv module.longshot.module.longshot-footlocker.aws_iam_user_policy.s3_policy module.longshot.module.longshot-footlocker.module.storage.aws_iam_user_policy.s3_policy
terraform state mv module.longshot.module.longshot-footlocker.aws_iam_user_policy.sqs_policy module.longshot.module.longshot-footlocker.module.queue.aws_iam_user_policy.sqs_policy
terraform state mv module.longshot.module.longshot-footlocker.aws_s3_bucket.storage module.longshot.module.longshot-footlocker.module.storage.aws_s3_bucket.bucket
terraform state mv module.longshot.module.longshot-footlocker.aws_sqs_queue.queue module.longshot.module.longshot-footlocker.module.queue.aws_sqs_queue.queue
terraform state mv module.longshot.module.longshot-footlocker.heroku_addon.redis module.longshot.module.longshot-footlocker.module.app.heroku_addon.redis
terraform state mv module.longshot.module.longshot-footlocker.heroku_app.app module.longshot.module.longshot-footlocker.module.app.heroku_app.app
terraform state mv module.longshot.module.longshot-footlocker.heroku_domain.domain module.longshot.module.longshot-footlocker.module.app.heroku_domain.domain
terraform state mv module.longshot.module.longshot-footlocker.heroku_drain.papertrail module.longshot.module.longshot-footlocker.module.app.heroku_drain.papertrail
terraform state mv module.longshot.module.longshot-footlocker.heroku_formation.queue module.longshot.module.longshot-footlocker.module.app.heroku_formation.queue
terraform state mv module.longshot.module.longshot-footlocker.heroku_formation.web module.longshot.module.longshot-footlocker.module.app.heroku_formation.web
terraform state mv module.longshot.module.longshot-footlocker.heroku_pipeline_coupling.app module.longshot.module.longshot-footlocker.module.app.heroku_pipeline_coupling.app
terraform state mv module.longshot.module.longshot-qa.aws_db_instance.database module.longshot.module.longshot-qa.module.database.aws_db_instance.database
terraform state mv module.longshot.module.longshot-qa.aws_iam_access_key.aws_key module.longshot.module.longshot-qa.module.iam_user.aws_iam_access_key.aws_key
terraform state mv module.longshot.module.longshot-qa.aws_iam_user.aws_user module.longshot.module.longshot-qa.module.iam_user.aws_iam_user.aws_user
terraform state mv module.longshot.module.longshot-qa.aws_iam_user_policy.s3_policy module.longshot.module.longshot-qa.module.storage.aws_iam_user_policy.s3_policy
terraform state mv module.longshot.module.longshot-qa.aws_iam_user_policy.sqs_policy module.longshot.module.longshot-qa.module.queue.aws_iam_user_policy.sqs_policy
terraform state mv module.longshot.module.longshot-qa.aws_s3_bucket.storage module.longshot.module.longshot-qa.module.storage.aws_s3_bucket.bucket
terraform state mv module.longshot.module.longshot-qa.aws_sqs_queue.queue module.longshot.module.longshot-qa.module.queue.aws_sqs_queue.queue
terraform state mv module.longshot.module.longshot-qa.heroku_addon.redis module.longshot.module.longshot-qa.module.app.heroku_addon.redis
terraform state mv module.longshot.module.longshot-qa.heroku_app.app module.longshot.module.longshot-qa.module.app.heroku_app.app
terraform state mv module.longshot.module.longshot-qa.heroku_domain.domain module.longshot.module.longshot-qa.module.app.heroku_domain.domain
terraform state mv module.longshot.module.longshot-qa.heroku_drain.papertrail module.longshot.module.longshot-qa.module.app.heroku_drain.papertrail
terraform state mv module.longshot.module.longshot-qa.heroku_formation.queue module.longshot.module.longshot-qa.module.app.heroku_formation.queue
terraform state mv module.longshot.module.longshot-qa.heroku_formation.web module.longshot.module.longshot-qa.module.app.heroku_formation.web
terraform state mv module.longshot.module.longshot-qa.heroku_pipeline_coupling.app module.longshot.module.longshot-qa.module.app.heroku_pipeline_coupling.app
```

</details>

**Comment by DFurnes** [comment]
Better yet, the Longshot application module now defaults to New Relic being enabled based on environment, but allows an optional `with_newrelic` override (so we can "force" disabling this for now on the inactive production `footlocker-internal` app). 1048629

The new plan, once this pull request is merged:

```
No changes
```

😌

**Comment by DFurnes** [comment]
Made a few tiny readability changes in a80188b, 59c0104, and 83c0ad1. As before, ez plan:

```
No changes
```

File: shared/rds_instance/main.tf [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  username = "${data.aws_ssm_parameter.database_username.value}"
+  password = "${data.aws_ssm_parameter.database_password.value}"
+
+  # TODO: We should migrate our account out of EC2-Classic, create
+  # a default VPC, and let resources be created in there by default!
+  db_subnet_group_name = "${var.subnet_group}"
+
+  vpc_security_group_ids = "${var.security_groups}"
+  publicly_accessible    = true
+
```
Comment (sheyd) [inline shared/rds_instance/main.tf]: Did we want to add `deletion_protection  = true` as a default value, or keep as `false`, and specify `true` manually for Prod instances?

**Comment by sheyd** [comment]
@DFurnes Added one question referring to https://github.com/DoSomething/infrastructure/issues/66 which we briefly discussed the other day. Otherwise, 🚢 !

**Comment by sheyd** [comment]
This is so good!!

File: shared/rds_instance/main.tf (hunk 1 above)
Comment (DFurnes) [reply shared/rds_instance/main.tf]: Ah yes!! Let me add that in a follow-up PR so this one stays as a no-op.

**Pull request merged**